README.rst
===================================================
Community Detection CliXO
===================================================

This repository creates a CDAPS compatible community detection Docker image using CliXO
packaged from https://github.com/fanzheng10/CliXO

Original paper: Kramer M, Dutkowski J, Yu M, Bafna V, Ideker T. Inferring gene ontologies from pairwise similarity data. Bioinformatics, 30: i34-i42. 2014. doi: 10.1093/bioinformatics/btu282

* A classical method that iteratively detects hierarchy based on cliques from changing edge weight threshold.
* CliXO only works on graph with edge weights.
* For reasonable runtime, the maximum size of network is of hundreds of nodes.

* Parameter `alpha`: threshold between cluster layers (default `0.1`)
* Parameter `beta`: merge similarity for overlapping clusters (default `0.5`)

Dependencies
------------

* `Docker <https://www.docker.com/>`_
* `make <https://www.gnu.org/software/make/>`_ (to build)
* Python (to build)

Direct invocation
------------------

Version `0.1.0` can be directly pulled from `Dockerhub <https://hub.docker.com/>`_ with this command:

.. code-block::

   docker pull coleslawndex/cdclixo:0.1.0

Building
--------

.. code-block::

   git clone https://github.com/idekerlab/cdclixo
   cd cdclixo
   make dockerbuild

Run **make** command with no arguments to see other build/deploy options including creation of Docker image

.. code-block::

   make

Output:

.. code-block::

   clean                remove all build, test, coverage and Python artifacts
   clean-build          remove build artifacts
   clean-pyc            remove Python file artifacts
   clean-test           remove test and coverage artifacts
   lint                 check style with flake8
   test                 run tests quickly with the default Python
   test-all             run tests on every Python version with tox
   coverage             check code coverage quickly with the default Python
   docs                 generate Sphinx HTML documentation, including API docs
   servedocs            compile the docs watching for changes
   testrelease          package and upload a TEST release
   release              package and upload a release
   dist                 builds source and wheel package
   install              install the package to the active Python's site-packages
   dockerbuild          build docker image and store in local repository
   dockerpush           push image to dockerhub


Usage
-----

.. code-block::

   docker run -v coleslawndex/cdclixo:0.1.0 -h

Credits
---------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

1.0.0b1 (2020-07-28)
---------------------

* Beta release of CliXO-1.0

* See https://github.com/fanzheng10/CliXO-1.0 for changes

0.1.0 (2020-01-06)
------------------

* First release


tests/test_cdclixo.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_cdclixo
----------------------------------

Tests for `cdclixo` module.
"""

import os
import sys
import unittest
import tempfile
import shutil


from cdclixo import cdclixocmd


class TestCdclixo(unittest.TestCase):

    def setUp(self):
        pass

    def tearDown(self):
        pass

    def test_parse_args(self):
        myargs = ['inputarg']
        res = cdclixocmd._parse_arguments('desc', myargs)
        self.assertEqual('inputarg', res.input)
        self.assertEqual(0.1, res.alpha)
        self.assertEqual(0.5, res.beta)

    def test_run_clixo_no_file(self):
        temp_dir = tempfile.mkdtemp()
        try:
            tfile = os.path.join(temp_dir, 'foo')
            myargs = [tfile]
            theargs = cdclixocmd._parse_arguments('desc', myargs)
            res = cdclixocmd.run_clixo('clixo', tfile, theargs)
            self.assertEqual(3, res)
        finally:
            shutil.rmtree(temp_dir)

    def test_run_clixo_empty_file(self):
        temp_dir = tempfile.mkdtemp()
        try:
            tfile = os.path.join(temp_dir, 'foo')
            open(tfile, 'a').close()
            myargs = [tfile]
            theargs = cdclixocmd._parse_arguments('desc', myargs)
            res = cdclixocmd.run_clixo('clixo', tfile, theargs)
            self.assertEqual(4, res)
        finally:
            shutil.rmtree(temp_dir)

    def test_main_invalid_file(self):
        temp_dir = tempfile.mkdtemp()
        try:
            tfile = os.path.join(temp_dir, 'foo')
            myargs = ['prog', tfile]
            res = cdclixocmd.main(myargs)
            self.assertEqual(3, res)
        finally:
            shutil.rmtree(temp_dir)

    def test_main_empty_file(self):
        temp_dir = tempfile.mkdtemp()
        try:
            tfile = os.path.join(temp_dir, 'foo')
            open(tfile, 'a').close()
            myargs = ['prog', tfile]
            res = cdclixocmd.main(myargs)
            self.assertEqual(4, res)
        finally:
            shutil.rmtree(temp_dir)


if __name__ == '__main__':
    sys.exit(unittest.main())


travis_pypi_setup.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Update encrypted deploy password in Travis config file
"""


from __future__ import print_function
import base64
import json
import os
from getpass import getpass
import yaml
from cryptography.hazmat.primitives.serialization import load_pem_public_key
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric.padding import PKCS1v15


try:
    from urllib import urlopen
except:
    from urllib.request import urlopen


GITHUB_REPO = 'idekerlab/cdclixo'
TRAVIS_CONFIG_FILE = os.path.join(
    os.path.dirname(os.path.abspath(__file__)), '.travis.yml')


def load_key(pubkey):
    """Load public RSA key, with work-around for keys using
    incorrect header/footer format.

    Read more about RSA encryption with cryptography:
    https://cryptography.io/latest/hazmat/primitives/asymmetric/rsa/
    """
    try:
        return load_pem_public_key(pubkey.encode(), default_backend())
    except ValueError:
        # workaround for https://github.com/travis-ci/travis-api/issues/196
        pubkey = pubkey.replace('BEGIN RSA', 'BEGIN').replace('END RSA', 'END')
        return load_pem_public_key(pubkey.encode(), default_backend())


def encrypt(pubkey, password):
    """Encrypt password using given RSA public key and encode it with base64.

    The encrypted password can only be decrypted by someone with the
    private key (in this case, only Travis).
    """
    key = load_key(pubkey)
    encrypted_password = key.encrypt(password, PKCS1v15())
    return base64.b64encode(encrypted_password)


def fetch_public_key(repo):
    """Download RSA public key Travis will use for this repo.

    Travis API docs: http://docs.travis-ci.com/api/#repository-keys
    """
    keyurl = 'https://api.travis-ci.org/repos/{0}/key'.format(repo)
    data = json.loads(urlopen(keyurl).read().decode())
    if 'key' not in data:
        errmsg = "Could not find public key for repo: {}.\n".format(repo)
        errmsg += "Have you already added your GitHub repo to Travis?"
        raise ValueError(errmsg)
    return data['key']


def prepend_line(filepath, line):
    """Rewrite a file adding a line to its beginning.
    """
    with open(filepath) as f:
        lines = f.readlines()

    lines.insert(0, line)

    with open(filepath, 'w') as f:
        f.writelines(lines)


def load_yaml_config(filepath):
    with open(filepath) as f:
        return yaml.load(f)


def save_yaml_config(filepath, config):
    with open(filepath, 'w') as f:
        yaml.dump(config, f, default_flow_style=False)


def update_travis_deploy_password(encrypted_password):
    """Update the deploy section of the .travis.yml file
    to use the given encrypted password.
    """
    config = load_yaml_config(TRAVIS_CONFIG_FILE)

    config['deploy']['password'] = dict(secure=encrypted_password)

    save_yaml_config(TRAVIS_CONFIG_FILE, config)

    line = ('# This file was autogenerated and will overwrite'
            ' each time you run travis_pypi_setup.py\n')
    prepend_line(TRAVIS_CONFIG_FILE, line)


def main(args):
    public_key = fetch_public_key(args.repo)
    password = args.password or getpass('PyPI password: ')
    update_travis_deploy_password(encrypt(public_key, password.encode()))
    print("Wrote encrypted password to .travis.yml -- you're ready to deploy")


if '__main__' == __name__:
    import argparse
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--repo', default=GITHUB_REPO,
                        help='GitHub repo (default: %s)' % GITHUB_REPO)
    parser.add_argument('--password',
                        help='PyPI password (will prompt if not provided)')

    args = parser.parse_args()
    main(args)


cdclixo/cdclixocmd.py
#!/usr/bin/env python

import os
import sys
import argparse
import subprocess


def _parse_arguments(desc, args):
    """
    Parses command line arguments
    :param desc:
    :param args:
    :return:
    """
    help_fm = argparse.RawDescriptionHelpFormatter
    parser = argparse.ArgumentParser(description=desc,
                                     formatter_class=help_fm)
    parser.add_argument('input',
                        help='Edge file in tab delimited format of '
                             'node1 node2 edgeWeight')

    parser.add_argument('--alpha', default=0.1, type=float,
                        help='Sets alpha parameter')
    parser.add_argument('--beta', default=0.5, type=float,
                        help='Sets beta parameter')
    parser.add_argument('--clixocmd', default='/clixo/clixo',
                        help='Full path to CliXO command')
    return parser.parse_args(args)


def run_clixo_cmd(args):
    """
    Runs docker
    :param cmd_to_run: command to run as list
    :return:
    """

    p = subprocess.Popen(args,
                         stdout=subprocess.PIPE,
                         stderr=subprocess.PIPE)

    out, err = p.communicate()

    return p.returncode, out, err


def run_clixo(clixocmd, graph, alpha=0.1, beta=0.5):

    """
    :outdir: the output directory to comprehend the output link file
    :param graph: input file
    :param overlap: bool, whether to enable overlapping community detection
    :param directed
    :return
    """
    if not os.path.isfile(graph):
        sys.stderr.write('No input file found\n')
        return 3

    if os.path.getsize(graph) == 0:
        sys.stderr.write('Input file is empty\n')
        return 4

    cmdargs = [clixocmd, '-i', graph, '-a', str(alpha),
               '-b', str(beta)]

    cmdecode, cmdout, cmderr = run_clixo_cmd(cmdargs)

    if cmdecode != 1:
        sys.stderr.write('Command failed with non-zero exit code: ' +
                         str(cmdecode) + ' : ' + str(cmderr) + '\n')
        return 1

    termtype = 'c-c'
    lines = cmdout.decode('utf-8').splitlines()
    for line in lines:
        if '#' == line[0]:
            continue
        elts = line.split()
        termcol = elts[2]
        if termcol == 'gene' and termtype == 'c-c':
            termtype = 'c-m'
        if termcol == 'default' and termtype == 'c-m':
            termtype = 'c-c'
        sys.stdout.write(elts[0] + ',' + elts[1] + ',' + termtype + ';')
    sys.stdout.flush()
    return 0


def main(args):
    """
    Main entry point for program
    :param args: command line arguments usually :py:const:`sys.argv`
    :return: 0 for success otherwise failure
    :rtype: int
    """
    desc = """
    Runs clixo on command line, sending output to standard
    out 
    """

    theargs = _parse_arguments(desc, args[1:])

    try:
        inputfile = os.path.abspath(theargs.input)

        return run_clixo(theargs.clixocmd, inputfile, alpha=theargs.alpha,
                         beta=theargs.beta)
    except Exception as e:
        sys.stderr.write('Caught exception: ' + str(e))
        return 2


if __name__ == '__main__':  # pragma: no cover
    sys.exit(main(sys.argv))